src/Workspaces/Core/Desktop/Workspace/Host/TemporaryStorage/TemporaryStorageServiceFactory.MemoryMappedInfo.cs

```diff
@@ -115,85 +106,91 @@ public Stream CreateWritableStream()
                 }
             }
 
-            private void StreamDisposed()
+            public void Dispose()
             {
-                lock (_memoryMappedFile)
+                Dispose(true);
+                GC.SuppressFinalize(this);
+            }
+
+            private void Dispose(bool disposing)
+            {
+                if (disposing)
                 {
-                    _streamCount--;
-                    if (_streamCount == 0 && _accessor != null)
+                    lock (_memoryMappedFile)
                     {
-                        _accessor.Dispose();
-                        _accessor = null;
+                        if (_accessor != null)
+                        {
+                            // dispose accessor it owns.
+                            // if someone explicitly called Dispose when streams given out are not
+                            // disposed yet, the accessor each stream has will simply stop working.
+                            //
+                            // it is caller's responsibility to make sure all streams it got from
+                            // the temporary storage are disposed before calling dispose on the storage.
+                            //
+                            // otherwise, finalizer will take care of disposing stuff as we used to be.
+                            _accessor.Dispose();
+                            _accessor = null;
+                        }
                     }
+
+                    // Dispose the memoryMappedFile
+                    _memoryMappedFile.Dispose();
                 }
             }
 
-            ~MemoryMappedInfo()
+            public static string CreateUniqueName(long size)
             {
-                Dispose(false);
+                return "Roslyn Temp Storage " + size.ToString() + " " + Guid.NewGuid().ToString("N");
             }
 
-            public void Dispose()
+            private unsafe sealed class CopiedMemoryMappedViewHandle : CriticalHandleZeroOrMinusOneIsInvalid
             {
-                Dispose(true);
-                GC.SuppressFinalize(this);
-            }
+                private readonly SafeMemoryMappedViewHandle _viewHandle;
 
-            private void Dispose(bool disposing)
-            {
-                lock (_memoryMappedFile)
+                public CopiedMemoryMappedViewHandle(SafeMemoryMappedViewHandle viewHandle)
                 {
-                    if (_accessor != null)
+                    _viewHandle = viewHandle;
+
+                    RuntimeHelpers.PrepareConstrainedRegions();
```

Comment:
Why would it be invalid?

---
In reply to: [116652491](https://github.com/dotnet/roslyn/pull/19522#discussion_r116652491) [](ancestors = 116652491)